visions/accessors.md

```diff
@@ -0,0 +1,348 @@
+
+A `read` or `modify` accessor can usually avoid a copy, but not always:  In practice, the implementation cannot delay the bottom half of the coroutine indefinitely.  This can lead to situations where the value is still needed after the coroutine ends, which requires copying the value from the coroutine’s execution frame into the calling context.  When this happens, a `_read` or `_modify` accessor is generally slower than a regular `get`, since it incurs copy overhead similar to a `get` in addition to the coroutine execution overhead.  This also limits the use of `read` or `modify` with property values that are not copyable.
+
+### Exposing transformed temporary values
+
+The `read` and `modify` accessors provide the ability to expose a transformed version of a stored value. This is important for dictionary mutation, which in the current API exposes the value as an Optional which is `nil` if the value is not currently set.  This allows a dictionary update such as `dict[key]?.modify()` to copy the dictionary value into a constructed optional in the top half of the accessor, then run the method call directly on the value, then run the bottom half to deconstruct the optional and store the result back into the dictionary storage.
+
+Another example:  Providing a `Span` over the contents of a String will require that we somehow handle the case where a short String is stored inline.
+
+### Access Scope
+
+The `get`/`set` accessors conceptually represent “instantaneous” access of the value.  After the `get` accessor returns, there is no relation between the returned copy of the property value and the containing value.  Each of the other accessors implies that the containing value must continue to exist for some period.  This period is generally fairly short, but optimization can extend this interval in order to simplify other operations:
+
+* `unsafe*Address` accessors return pointers into the containing value.  This is safe for the caller of these accessors because the compiler knows about this relationship and can extend the lifetime of the containing value as needed.  (These accessors are nominally “unsafe” because their implementation requires constructing an unsafe pointer and there are no checks to ensure that the pointer so constructed is in fact valid.)
+* `read`/`modify` expose a value for the lifetime of a coroutine. Coroutines enforce that the containing value remains alive for the duration of the coroutine.  Note that in current Swift, the coroutines are generally quite short-lived and the compiler copies the value into or out of the coroutine fairly aggressively.  In the future, the compiler will likely become more adept at expanding the coroutine lifetime to reduce such copying.
+* `borrow`/`mutate` return a borrow of the property value.  This borrow has an implied dependence on the containing value, and the compiler must guarantee that the containing value outlives the property access.
```

Comment:
I imagine that borrow/mutate would not work for the example I gave, given that the accessor would need to take a strong reference and there wouldn't be a "bottom half coroutine" to drop it?

Reply:
Unless `unsafe*Address`, the `borrow`/`mutate` accessors should be amenable to various checks at the time the accessor is compiled.  For example, we could warn or error if you tried to return an interior pointer to a class object where there was not an associated strong reference to keep it alive.  So `borrow`/`mutate` would be safe in that we should be able to issue an error if you try to do this -- but for your example as written, it would likely require some variation of coroutine-based accessor.